README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# numberwang <img src="man/figures/logo.png" align="right" height="230/"/>

<!-- badges: start -->

![](https://img.shields.io/badge/cool-useless-green.svg)
![](https://img.shields.io/badge/test%20coverage-100%25-blue.svg)
[![R-CMD-check](https://github.com/coolbutuseless/numberwang/workflows/R-CMD-check/badge.svg)](https://github.com/coolbutuseless/numberwang/actions)
<!-- badges: end -->

`numberwang` will convert floating point numbers (and integers) to their
word representations, and vice versa.

The key differentiator of this package, compared to
[{nombre}](https://cran.r-project.org/package=nombre), is that it
supports decimal representations by listing individual decimal digits.

See below for a comparison with `{nombre}`, `{english}`, `{spanish}` and
`{xfun}`.

## What’s in the box

-   `num_to_words(num, decimals = 3)` converts numeric vectors to
    character vectors
-   `words_to_num(words)` converts character vectors to numeric vectors

## Installation

You can install from
[GitHub](https://github.com/coolbutuseless/numberwang) with:

``` r
# install.package('remotes')
remotes::install_github('coolbutuseless/numberwang')
```

## That’s `numberwang`

<img src="man/figures/thats-numberwang.jpg" />

``` r
library(numberwang)

c(12, pi, exp(10))
```

    #> [1]    12.000000     3.141593 22026.465795

``` r
num_to_words(c(12, pi, exp(10)))
```

    #> [1] "twelve"                                               
    #> [2] "three point one four two"                             
    #> [3] "twenty-two thousand and twenty-six point four six six"

``` r
words_to_num(c(
  'negative fifteen point seven',
  "twenty-three point seven zero one",
  "ten million, two thousand and forty two"
))
```

    #> [1]      -15.700       23.701 10002042.000

## Limitations

-   Floating point representation isn’t exact, which means that for very
    large numbers or very small numbers, and for lots of numbers
    inbetween, all the digits after the first few must often be taken
    with a grain of salt.
-   E.g.
    `formatC(0.3-0.2, format = 'f', digits = 20) = 0.09999999999999997780`
-   E.g. `1e25` should just be a `1` followed by zeros, but in floating
    point representation we get:
    `formatC(1e25, format = 'f', digits = 0) = 10000000000000000905969664`
-   The `{english}` and `{nombre}` packages will warn you about
    precision loss like this, but `{numberwang}` won’t.

## Feature comparison

|                  | numberwang     | nombre                  | english  | spanish   | xfun     |
|------------------|----------------|-------------------------|----------|-----------|----------|
| function         | num\_to\_words | nom\_card               | english  | to\_words | n2w      |
| working range    | double()       | double()                | +/- 2^90 | \[0,999\] | +/- 1e15 |
| decimals         | Yes            | as fractions            | No       | No        | No       |
| spanish          | No             | No                      | No       | Yes       | No       |
| words-to-numbers | words\_to\_num | nom\_uncard (ints only) | No       | No        | No       |

## Just for fun - Let’s rotate the board!

<img src="man/figures/lets-rotate-the-board.jpg" />

``` r
exp(300)
```

    #> [1] 1.942426e+130

``` r
formatC(
  exp(300), format = 'f', digits = 0
)
```

    #> [1] "19424263952412558251527551342607068708348876681949259689720520473778975990853476787505520791207218561168735878224980403355291484160"

``` r
num_to_words(exp(300))
```

      nineteen duoquadragintillion, four hundred and twenty-four
      unquadragintillion, two hundred and sixty-three quadragintillion, nine
      hundred and fifty-two novemtrigintillion, four hundred and twelve
      octatrigintillion, five hundred and fifty-eight septentrigintillion,
      two hundred and fifty-one sextrigintillion, five hundred and
      twenty-seven quintrigintillion, five hundred and fifty-one
      quattuortrigintillion, three hundred and forty-two tretrigintillion,
      six hundred and seven duotrigintillion, sixty-eight untrigintillion,
      seven hundred and eight trigintillion, three hundred and forty-eight
      novemvigintillion, eight hundred and seventy-six octavigintillion, six
      hundred and eighty-one septenvigintillion, nine hundred and forty-nine
      sexvigintillion, two hundred and fifty-nine quinvigintillion, six
      hundred and eighty-nine quattuorvigintillion, seven hundred and twenty
      trevigintillion, five hundred and twenty duovigintillion, four hundred
      and seventy-three unvigintillion, seven hundred and seventy-eight
      vigintillion, nine hundred and seventy-five novemdecillion, nine
      hundred and ninety octadecillion, eight hundred and fifty-three
      septendecillion, four hundred and seventy-six sexdecillion, seven
      hundred and eighty-seven quindecillion, five hundred and five
      quattuordecillion, five hundred and twenty tredecillion, seven hundred
      and ninety-one duodecillion, two hundred and seven undecillion, two
      hundred and eighteen decillion, five hundred and sixty-one nonillion,
      one hundred and sixty-eight octillion, seven hundred and thirty-five
      septillion, eight hundred and seventy-eight sextillion, two hundred and
      twenty-four quintillion, nine hundred and eighty quadrillion, four
      hundred and three trillion, three hundred and fifty-five billion, two
      hundred and ninety-one million, four hundred and eighty-four thousand,
      one hundred and sixty

## Related Software

-   [nombre](https://cran.r-project.org/package=nombre)
-   [english](https://cran.r-project.org/package=english)
-   [spanish](https://cran.r-project.org/package=spanish)
-   [xfun](https://cran.r-project.org/package=xfun)

## Acknowledgements

-   R Core for developing and maintaining the language.
-   CRAN maintainers, for patiently shepherding packages onto CRAN and
    maintaining the repository


NEWS.md
# numberwang 0.1.0

* Initial release   2021-08-29


tests/testthat.R
library(testthat)
library(numberwang)

test_check("numberwang")


tests/testthat/test-small-int.R


test_that("small-int-to-words works", {
  expect_equal(
    small_int_to_words(603),
    "six hundred and three"
  )


  expect_equal(
    small_int_to_words(0),
    ""
  )

  expect_equal(
    small_int_to_words(12),
    "twelve"
  )

  expect_equal(
    small_int_to_words(123),
    "one hundred and twenty-three"
  )

  expect_equal(
    small_int_to_words(110),
    "one hundred and ten"
  )

  expect_equal(
    small_int_to_words(101),
    "one hundred and one"
  )

  expect_equal(
    small_int_to_words(100),
    "one hundred"
  )

  expect_equal(
    small_int_to_words(120),
    "one hundred and twenty"
  )

})


tests/testthat/test-stress-test.R



test_that("stress test", {

  set.seed(1)
  x <- c(
    1:1000,
    999:1222,
    sample(1e7, 1000)
  )

  w <- num_to_words(x)
  out <- words_to_num(w)

  expect_equal(x, out)

})


R/lex.R


# nocov start

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Break a string into labelled tokens based upon a set of patterns
#'
#' @param text a single character string
#' @param regexes a named vector of regex strings.  Each string represents
#'                a regex to match a token, and the name of the string is the
#'                label for the token.  Each regex can contain an explicit
#'                captured group using the standard \code{()} brackets. If a regex
#'                doesn't not define a captured group then the entire regex will
#'                be captured. The regexes
#'                will be processed in order such that an early match takes
#'                 precedence over any later match.
#' @param verbose print more information about the matching process. default: FALSE
#'
#' @return a named character vector with the names representing the token type
#'         with the value being the element extracted by the corresponding
#'         regular expression.
#'
#' @examples
#' \dontrun{
#' lex("hello there 123.45", regexes=c(number=re$number, word="(\\w+)", whitespace="(\\s+)"))
#' }
#' @import stringi
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lex <- function(text, regexes, verbose=FALSE) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # disallow multiple capture groups in a single pattern.
  # i.e. regexes = c("(a|b)", "(c)|(d)")
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  captured_groups <- stringi::stri_match_all(regexes, regex = "(?<!\\\\)\\([^?]")

  n_captured_groups <- vapply(captured_groups, FUN = function(x) {
    if (anyNA(x)) {
      0L
    } else {
      nrow(x)
    }
  }, integer(1))
  if (any(n_captured_groups > 1)) {
    stop("Regexes can define at most only a single capture group. Patterns which need fixing",
         deparse(regexes[n_captured_groups > 1]))
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Any regex that has 0 capture groups has its whole regex become the
  # capture group
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  idx <- n_captured_groups == 0
  regexes[idx] <- paste0("(", regexes[idx], ")")

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Insert a default pattern to match anything missed by the provided regexes
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  regexes        <- c(regexes, .missing="(.)")
  regex_labels   <- names(regexes)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # All regexes must be named
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  stopifnot(!anyNA(regex_labels))
  stopifnot(!any(regex_labels == ''))

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Combine all the patterns into a single regex
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  regex <- paste(regexes, collapse='|')

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Match all regex against the text
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  token_matching  <- stringi::stri_match_all(text, regex = regex)[[1]]

  if (verbose) {
    colnames(token_matching) <- c("all", regex_labels)
    print(token_matching)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Extract the actual token and the regex which matched the token
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  regex_idx      <- apply(token_matching[, -1, drop=FALSE], 1, function(x) {  which(!is.na(x))})
  tokens         <- apply(token_matching[, -1, drop=FALSE], 1, function(x) {x[which(!is.na(x))]})

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # if 'regex_idx' is a list, then a location was matched by multiple regexes
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (is.list(regex_idx)) {
    lens <- lengths(regex_idx)
    idx  <- which(lens > 1)
    stop("lex issues at the following locations within 'text': ", deparse(idx))
  }


  names(tokens)  <- regex_labels[regex_idx]

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # If any tokens were captured by the '.missing' regex, then show
  # a warning message
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (verbose && any(names(tokens) == '.missing')) {
    not_captured <- sort(unique(tokens[names(tokens) == '.missing']))
    warning("The following characters were not captured: ", deparse(not_captured))
  }

  tokens
}



# nocov end


R/words-to-num.R



num_patterns <- c(
  hundred    = "\\w*? hundred",
  lion       = "thousand|\\w+llion",
  ten        = "ten|twenty|thirty|forty|fifty|sixty|seventy|eighty|ninety",
  teen       = "eleven|twelve|thirteen|fourteen|fifteen|sixteen|seventeen|eighteen|nineteen",
  digit      = "zero|one|two|three|four|five|six|seven|eight|nine",
  point      = "point|\\.",
  negative   = "negative",
  other      = "[^\\s]+",
  whitespace = '\\s+'
)


hundreds_num <- c(
  `one hundred`   = 100,
  `two hundred`   = 200,
  `three hundred` = 300,
  `four hundred`  = 400,
  `five hundred`  = 500,
  `six hundred`   = 600,
  `seven hundred` = 700,
  `eight hundred` = 800,
  `nine hundred`  = 900
)

tens_num <- c(
  `ten`     = 10,
  `twenty`  = 20,
  `thirty`  = 30,
  `forty`   = 40,
  `fifty`   = 50,
  `sixty`   = 60,
  `seventy` = 70,
  `eighty`  = 80,
  `ninety`  = 90
)

digits_num <- c(
  `zero`  = 0,
  `one`   = 1,
  `two`   = 2,
  `three` = 3,
  `four`  = 4,
  `five`  = 5,
  `six`   = 6,
  `seven` = 7,
  `eight` = 8,
  `nine`  = 9
)

teens_num <- c(
  `eleven`    = 11,
  `twelve`    = 12,
  `thirteen`  = 13,
  `fourteen`  = 14,
  `fifteen`   = 15,
  `sixteen`   = 16,
  `seventeen` = 17,
  `eighteen`  = 18,
  `nineteen`  = 19
)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a single character string into a numeric value
#'
#' @param words single character string
#'
#'
#' @examples
#' \dontrun{
#' words_to_num_single("twelve")
#' }
#'
#' @return numeric value
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
words_to_num_single <- function(words) {
  stopifnot(length(words) == 1)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Get rid of non number stuff
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  txt <- words
  txt <- gsub("-|,", " ", txt)
  txt <- gsub(" and", "", txt)
  txt <- gsub("\\s+", " ", txt)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # split into tokens
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  tokens <- lex(txt, num_patterns)
  neg    <- 'negative' %in% names(tokens)
  tokens <- tokens[!names(tokens) %in% c('whitespace', 'negative')]

  if (any(names(tokens) == 'other')) {
    stop("Unknown symbols in text: ", deparse(unname(tokens[names(tokens) == 'other'])))
  }


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Iterate over tokens, calculating the total numeric value
  #
  # * Assimulate values until we find one of the "-illion" markers.
  # * When a "-illion" marker is hit, work out its power-of-ten from its index
  #    in the 'big' vector, then multiple by the accumulator and add to total.
  #    Then reset the accumlator to zero
  #
  # This is going to have issues with floating-point accuracy, and would
  # possibly be more correctly done by constructing a character representation
  # of the number and then using "as.numeric()".
  #
  # But this is good enough for now.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  total <- 0
  accum <- 0
  for (i in seq_along(tokens)) {
    type <- names(tokens)[i]
    val  <- tokens[[i]]
    # message(sprintf("[%i/%i] %s %s", accum, total, type, val))
    if (type == 'lion') {
      idx <- which(big == val)
      mult  <- 10 ^ (3 * (idx - 1))
      total <- total + accum * mult
      accum <- 0
    } else if (type == 'hundred') {
      accum <- accum + hundreds_num[[val]]
    } else if (type == 'ten') {
      accum <- accum + tens_num[[val]]
    } else if (type == 'teen') {
      accum <- accum + teens_num[[val]]
    } else if (type == 'digit') {
      accum <- accum + digits_num[[val]]
    } else if (type == 'point') {
      break
    }
  }
  total <- total + accum

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Process any decimals after the decimal 'point'
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (length(tokens) > i) {
    # some decimal places to process
    N        <- length(tokens) - i
    decimals <- tokens[seq(i+1, length(tokens), 1)]
    vals     <- digits_num[decimals]
    frac     <- as.numeric(paste(vals, collapse="")) / 10^N
    total    <- total + frac
  }

  if (neg) {
    total <- -total
  }

  total
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a character strings into numeric values
#'
#' @param words character vector
#'
#' @return numeric vector
#'
#' @examples
#' \dontrun{
#' words_to_num(c("twelve", "one hundred point three"))
#' }
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
words_to_num <- function(words) {
  unname(
    vapply(words, words_to_num_single, double(1))
  )
}



if (FALSE) {
  words <- "one hundred million, two thousand and forty two"
  words_to_num_single(words)
}









R/num-to-words.R


ones <- list(
  `0` = NULL,
  `1` = 'one',
  `2` = 'two',
  `3` = 'three',
  `4` = 'four',
  `5` = 'five',
  `6` = 'six',
  `7` = 'seven',
  `8` = 'eight',
  `9` = 'nine'
)

frac_ones <- list(
  `0` = 'zero',
  `1` = 'one',
  `2` = 'two',
  `3` = 'three',
  `4` = 'four',
  `5` = 'five',
  `6` = 'six',
  `7` = 'seven',
  `8` = 'eight',
  `9` = 'nine'
)

teens <- list(
  `0` = 'ten',
  `1` = 'eleven',
  `2` = 'twelve',
  `3` = 'thirteen',
  `4` = 'fourteen',
  `5` = 'fifteen',
  `6` = 'sixteen',
  `7` = 'seventeen',
  `8` = 'eighteen',
  `9` = 'nineteen'
)

tens <- list(
  `0` = '',
  `1` = 'ten',
  `2` = 'twenty',
  `3` = 'thirty',
  `4` = 'forty',
  `5` = 'fifty',
  `6` = 'sixty',
  `7` = 'seventy',
  `8` = 'eighty',
  `9` = 'ninety'
)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# From: https://lcn2.github.io/mersenne-english-name/tenpower/tenpower.html
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
big <- c(
  "", "thousand", "million", "billion", "trillion", "quadrillion",
  "quintillion", "sextillion", "septillion", "octillion", "nonillion",
  "decillion", "undecillion", "duodecillion", "tredecillion", "quattuordecillion",
  "quindecillion", "sexdecillion", "septendecillion", "octadecillion",
  "novemdecillion", "vigintillion", "unvigintillion", "duovigintillion",
  "trevigintillion", "quattuorvigintillion", "quinvigintillion",
  "sexvigintillion", "septenvigintillion", "octavigintillion",
  "novemvigintillion", "trigintillion", "untrigintillion", "duotrigintillion",
  "tretrigintillion", "quattuortrigintillion", "quintrigintillion",
  "sextrigintillion", "septentrigintillion", "octatrigintillion",
  "novemtrigintillion", "quadragintillion", "unquadragintillion",
  "duoquadragintillion", "trequadragintillion", "quattuorquadragintillion",
  "quinquadragintillion", "sexquadragintillion", "septenquadragintillion",
  "octaquadragintillion", "novemquadragintillion", "quinquagintillion",
  "unquinquagintillion", "duoquinquagintillion", "trequinquagintillion",
  "quattuorquinquagintillion", "quinquinquagintillion", "sexquinquagintillion",
  "septenquinquagintillion", "octaquinquagintillion", "novemquinquagintillion",
  "sexagintillion", "unsexagintillion", "duosexagintillion", "tresexagintillion",
  "quattuorsexagintillion", "quinsexagintillion", "sexsexagintillion",
  "septensexagintillion", "octasexagintillion", "novemsexagintillion",
  "septuagintillion", "unseptuagintillion", "duoseptuagintillion",
  "treseptuagintillion", "quattuorseptuagintillion", "quinseptuagintillion",
  "sexseptuagintillion", "septenseptuagintillion", "octaseptuagintillion",
  "novemseptuagintillion", "octagintillion", "unoctogintillion",
  "duooctogintillion", "treoctogintillion", "quattuoroctogintillion",
  "quinoctogintillion", "sexoctogintillion", "septenoctogintillion",
  "octaoctogintillion", "novemoctogintillion", "nonagintillion",
  "unnonagintillion", "duononagintillion", "trenonagintillion",
  "quattuornonagintillion", "quinnonagintillion", "sexnonagintillion",
  "septennonagintillion", "octanonagintillion", "novemnonagintillion",
  "centillion", "cenuntillion", "cendotillion"
)



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' convert a number in range [0,999] to words
#'
#' @param num integer in range 0-999
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
small_int_to_words <- function(num) {

  num <- as.integer(num)
  stopifnot(num >= 0, num <= 999)

  x <- strsplit(as.character(num), "")[[1]]

  bits <- c()
  if (length(x) == 0 || all(x == '0')) {
    return("")
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # pad with "0" at front to make the logic a bit easier
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  while (length(x) != 3) {
    x <- c('0', x)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Hundreds
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (x[1] != '0') {
    bits <- c(bits, ones[[x[1]]], "hundred")
    if (x[2] != '0' || x[3] != '0') {
      bits <- c(bits, 'and')
    }
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # tens + ones
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (x[2] > "1") {
    this_tens <- tens[[x[2]]]
    this_ones <- ones[[x[3]]]
    if (x[3] != '0') {
      bits <- c(bits, paste(this_tens, this_ones, sep="-"))
    } else {
      bits <- c(bits, this_tens)
      bits <- c(bits, this_ones)
    }

  } else if (x[2] == '1') {
      bits <- c(bits, teens[[x[3]]])
  } else {
    bits <- c(bits, ones[[x[3]]])
  }


  paste(bits, collapse = " ")
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create the cache of words for all numbers in range [0,999]
# also include these numbers with zero-prefix.
# This will be used for faster lookup
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
int_cache <- vapply(0:999, small_int_to_words, character(1))



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' convert a character vector of triplets into a word description
#'
#' @param triplets characcter vector of number triplets
#'
#' @examples
#' \dontrun{
#' triplets_to_words(c('123', '345'))
#' }
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
triplets_to_words <- function(triplets) {
  bigs      <- rev(big[seq.int(length(triplets))])
  trips     <- int_cache[as.integer(triplets) + 1L]

  keep_idx <- trips != ""
  bigs     <- bigs [keep_idx]
  trips    <- trips[keep_idx]

  N <- length(trips)

  last_trip <- trips[N]
  if (!grepl("hundred", last_trip) && N > 1) {
    first <- paste(trips[-N], bigs[-N], sep = ' ', collapse = ", ")
    last  <- paste("and", last_trip)
    paste(first, last)
  } else {
    paste(trips, bigs, sep = ' ', collapse = ", ")
  }
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a sequence of digits to a sequence of words
#'
#' This function trims traling zeros and prefixes the result with the word "point"
#'
#' @param digits single character with just digits e.g. "1056"
#'
#' @return single character string e.g. "point one zero five six"
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
decimals_to_words <- function(digits) {
  # if (is.null(digits) || length(digits) == 0 || is.na(digits)) {
  #   return("")
  # }

  # Drop trailing zeros
  digits <- gsub("0+$", "", digits)
  if (digits == "") return("")

  digits <- strsplit(digits, "")[[1]]
  res <- paste(frac_ones[digits], collapse=" ")
  paste("point", res)
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert numbers to words
#'
#' @param num numeric vector
#' @param decimals number of decimal places
#'
#' @return character vector
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
num_to_words <- function(num, decimals = 3) {

  char <- formatC(abs(num), big.mark=",", format = 'f', digits = decimals)

  int_frac <- do.call(rbind, strsplit(char, "\\."))

  all_triplets <- strsplit(int_frac[,1], ",")
  ints  <- vapply(all_triplets, triplets_to_words, character(1))

  if (decimals == 0) {
    fracs <- NULL
  } else {
    fracs <- vapply(int_frac[,2], decimals_to_words, character(1))
  }

  res <- trimws(paste(ints, fracs))
  res <- gsub("\\s+", " ", res)
  res <- ifelse(num < 0, paste("negative", res), res)
  res[res == ""] <- 'zero'
  res
}




if (FALSE) {
  num <- c(12, 100, 101, 110, 111, 120, 121, 129, exp(15))
  num_to_words(num, decimals = 4)
  decimals <- 3

  x <- 10^(sample(30))
  bench::mark(
    nombre::nom_ord(x),
    num_to_words(x, decimals = 0),
    check = FALSE
  )

}








